indigo_api/data_migrations/akn3.py

```diff
@@ -0,0 +1,446 @@
+        # Note: if there are more than one schedules named e.g. "schedule" only the last one will be updated by this
+        doc.frbr_uri = doc.frbr_uri
+        self.basics(doc, mappings)
+        self.tables_blocklists(doc, mappings)
+        self.subsections_items_paras(doc, mappings)
+        self.term_ids(doc, mappings)
+        self.separators_eids(doc, mappings)
+        self.internal_references(doc, mappings)
+        self.annotation_anchors(mappings)
+
+        # just for tests
+        return mappings
+
+    def safe_update(self, element, mappings, old, new):
+        """ Updates ids and mappings:
+        - First, check `mappings` for `old`:
+            if it's already there, check that it maps to `new`. Log a warning if not.
+        - Then, rewrite_ids.
+        - If it wasn't already in, update `mappings`.
+        """
+        if old == new:
+            log.warning(f"Not updating unchanged id: {old}")
+            return
+
+        if mappings.get("old"):
+            if not mappings[old] == new:
+                log.warning(f"New id mapping differs from existing: {old} -> {mappings[old]}; ignoring {new}")
+            # rewrite ids but don't update mappings
+            rewrite_ids(element, old, new)
+            return
+
+        mappings.update(rewrite_ids(element, old, new))
+
+    def traverse_mappings(self, old, mappings):
+        if old in mappings:
+            old = self.traverse_mappings(mappings[old], mappings)
+
+        return old
+
+    def clean_number(self, num):
+        # e.g. 1.3.2. / (3) / (dA) / (12 bis)
+        return num.strip("().").replace(" ", "").replace(".", "-")
+
+    def get_parent_id(self, node):
+        parent = node.getparent()
+        if parent is not None:
+            parent_id = parent.get("id")
+            return parent_id if parent_id and not parent_id.startswith("att_") else self.get_parent_id(parent)
+
+    def paras_to_hcontainers(self, doc, mappings):
+        """ Update all instances un-numbered paragraphs to hcontainers. Slaw generates these when
+        a block of text is encountered in a hierarchical element.
+
+        This ALSO changes the id of the element to match AKN 3 styles, but using the id attribute (not eId)
+
+        <paragraph id="section-1.paragraph0">
+          <content>
+            <p>text</p>
+          </content>
+        </paragraph>
+
+        becomes
+
+        <hcontainer id="section-1.hcontainer_1">
+          <content>
+            <p>text</p>
+          </content>
+        </hcontainer>
+        """
+        for name, root in self.components(doc).items():
+            for para in root.xpath('.//a:paragraph[not(a:num)]', namespaces=self.nsmap):
+                para.tag = f'{{{doc.namespace}}}hcontainer'
+                # new id is based on the number of preceding hcontainer siblings
+                num = len(para.xpath('preceding-sibling::a:hcontainer', namespaces=self.nsmap)) + 1
+                old_id = para.get('id') or 'paragraph-1'
```

Comment:
added this for case when the para doesn't have an id

Reply:
maybe we should add a log message here, in case it happens somewhere other than in stubs?